packages/perseus/src/widgets/interactive-graphs/graphs/quadratic.tsx

```diff
@@ -0,0 +1,120 @@
+    const [a, b, c] = coeffRef.current;
+
+    // Calculate the y value based on the current x value
+    const y = (x) => (a * x + b) * x + c;
+
+    // Set the xy values for the Parametric plot
+    const xy = (x: number) => [x, y(x)] as vec.Vector2;
+
+    // Determine the range / domain of the graph
+    const t: vec.Vector2 = [range[0][0], range[0][1]] as vec.Vector2;
+
+    const handleOnMove = (destination: vec.Vector2, elementId: number) => {
+        // If the destination is invalid, we want do not want to move the point
+        const validDestination = isValidDestination(
+            destination,
+            elementId,
+            coords,
+        );
+        if (validDestination === false) {
+            return;
+        }
+
+        dispatch(movePoint(elementId, destination));
+    };
+
+    return (
+        <>
+            <Plot.Parametric xy={xy} t={t} color={color.blue} />
+            {coords.map((coord, i) => (
+                <StyledMovablePoint
+                    key={"point-" + i}
+                    point={coord}
+                    onMove={(destination) => handleOnMove(destination, i)}
+                />
+            ))}
+        </>
+    );
+}
+
+// Ensure that we are only snapping to coordinates that result in a valid quadratic equation
+export const isValidDestination = (
+    destination: vec.Vector2,
+    elementId: number,
+    coords: QuadraticCoords,
+): boolean => {
+    // Set up the new coords and check if the quadratic coefficients are valid
+    const newCoords: QuadraticCoords = [...coords];
+    newCoords[elementId] = destination;
+    const QuadraticCoefficients = getQuadraticCoefficients(newCoords);
+
+    // If the new destination results in an invalid quadratic equation, we don't want to move the point
+    if (QuadraticCoefficients === undefined) {
+        return false;
+    }
+
+    return true;
+};
+
+// Get the quadratic coefficients from the 3 control points
+export const getQuadraticCoefficients = (
+    coords: QuadraticCoords,
+): QuadraticCoefficient | undefined => {
+    const p1 = coords[0];
+    const p2 = coords[1];
+    const p3 = coords[2];
+
+    // If the denominator is 0, we are going to return undefined as we are
+    // unable to calculate the quadratic coefficients when they hit infinity
+    const denom = (p1[0] - p2[0]) * (p1[0] - p3[0]) * (p2[0] - p3[0]);
+    if (denom === 0) {
+        return;
+    }
+
+    const a =
+        (p3[0] * (p2[1] - p1[1]) +
+            p2[0] * (p1[1] - p3[1]) +
+            p1[0] * (p3[1] - p2[1])) /
+        denom;
+    const b =
+        (p3[0] * p3[0] * (p1[1] - p2[1]) +
+            p2[0] * p2[0] * (p3[1] - p1[1]) +
+            p1[0] * p1[0] * (p2[1] - p3[1])) /
+        denom;
+    const c =
+        (p2[0] * p3[0] * (p2[0] - p3[0]) * p1[1] +
+            p3[0] * p1[0] * (p3[0] - p1[0]) * p2[1] +
+            p1[0] * p2[0] * (p1[0] - p2[0]) * p3[1]) /
+        denom;
+    return [a, b, c];
```

Comment:
This seems extremely magical 😂 . Is this a widely known algorithm? It might be nice to include a comment linking to an explanation.

One thing that isn't immediately clear to me (okay, at all clear to me) from looking at this is whether the order of the provided `coords` matters. I believe it doesn't, but it would be cool to have a test demonstrating that.

Reply:
(We've discussed this elsewhere, but I figured it's good to keep all the data on the ticket in case we ever search back.) 

I added a comment and a test for this to confirm that the order does not matter! 

I found the source for the original logic, which was created 11 years ago (!) : https://phabricator.khanacademy.org/D2413

When I have some spare time, I'll try to fully break down the math so that I can provide a clearer description of what is going on. (Mostly as a self-curiosity)